x-pack/platform/plugins/shared/onechat/server/services/agents/modes/default/graph.ts

```diff
@@ -66,29 +76,69 @@ export const createAgentGraph = ({
     const messages = state.addedMessages;
     const lastMessage: AIMessage = messages[messages.length - 1];
     if (lastMessage && lastMessage.tool_calls?.length) {
+      const pendingCalls = lastMessage.tool_calls.length;
+      const current = state.toolCallCount;
+      const budget = state.maxToolCalls ?? maxToolCalls;
+      if (current + pendingCalls > budget) {
+        // budget exhausted -> finalize without executing these calls
+        return 'finalize';
+      }
       return 'tools';
     }
     return '__end__';
   };
 
   const toolHandler = async (state: StateType) => {
     const toolNodeResult = await toolNode.invoke(state.addedMessages);
+    const executedToolMessages = toolNodeResult.filter((m) => isToolMessage(m));
+    const increment = executedToolMessages.length;
 
     return {
       addedMessages: [...toolNodeResult],
+      toolCallCount: state.toolCallCount + increment,
     };
   };
 
+  const finalizeAnswer = async (state: StateType) => {
+    // Safety: handle last AI message if it attempted new tool calls we didn't execute.
+    const added: BaseMessage[] = [...state.addedMessages];
+    const last = added[added.length - 1] as AIMessage | undefined;
+    if (last && last.tool_calls && last.tool_calls.length > 0) {
+      const textOnly = extractTextContent(last).trim();
+      if (!textOnly) {
+        added.pop();
+      } else {
+        // Replace with a simple assistant tuple to avoid lingering provider metadata/tool_call chunks
+        (added as any)[added.length - 1] = ['assistant', textOnly];
+      }
```

Comment:
What are we trying to address here? we should always have a tool result as last message when we arrive in this step, right? 

```
-user message
[loop]
- assistant message
- tool message
[/loop]
<== we are always here, right?
```

So I'm not sure to get what this code is there for?